src/objectid.ts

```diff
@@ -43,54 +43,49 @@ export class ObjectId {
   /**
    * Create an ObjectId type
    *
-   * @param id - Can be a 24 character hex string, 12 byte binary Buffer, or a number.
+   * @param inputId - Can be a 24 character hex string, 12 byte binary Buffer, or a number.
    */
-  constructor(id?: string | Buffer | number | ObjectIdLike | ObjectId) {
-    if (!(this instanceof ObjectId)) return new ObjectId(id);
+  constructor(inputId?: string | Buffer | number | ObjectIdLike | ObjectId) {
+    if (!(this instanceof ObjectId)) return new ObjectId(inputId);
 
-    // Duck-typing to support ObjectId from different npm packages
-    if (id instanceof ObjectId) {
-      this[kId] = id.id;
-      this.__id = id.__id;
-    }
-
-    if (typeof id === 'object' && id && 'id' in id) {
-      if ('toHexString' in id && typeof id.toHexString === 'function') {
-        this[kId] = Buffer.from(id.toHexString(), 'hex');
+    let workingId;
+    if (typeof inputId === 'object' && inputId && 'id' in inputId) {
+      if (typeof inputId.id !== 'string' && !ArrayBuffer.isView(inputId.id)) {
+        throw new BSONTypeError(
+          'Argument passed in must have an id that is of type string or Buffer'
+        );
+      }
+      if ('toHexString' in inputId && typeof inputId.toHexString === 'function') {
+        workingId = Buffer.from(inputId.toHexString(), 'hex');
       } else {
-        this[kId] = typeof id.id === 'string' ? Buffer.from(id.id) : id.id;
+        workingId = inputId.id;
       }
+    } else {
+      workingId = inputId;
     }
-
-    // The most common use case (blank id, new objectId instance)
-    if (id == null || typeof id === 'number') {
+    if (workingId == null || typeof workingId === 'number') {
```

Comment:
nit: let's put a newline (and maybe a comment) here. At this stage the workingId should be setup, what follows is how the workingId gets used to make our ObjectId